aws/resource_aws_ec2_traffic_mirror_filter_rule.go

```diff
@@ -0,0 +1,370 @@
+							Optional: true,
+						},
+					},
+				},
+			},
+			"traffic_direction": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"ingress",
+					"egress",
+				}, false),
+			},
+		},
+	}
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	filterId := d.Get("traffic_mirror_filter_id")
+
+	input := &ec2.CreateTrafficMirrorFilterRuleInput{
+		TrafficMirrorFilterId: aws.String(filterId.(string)),
+		DestinationCidrBlock:  aws.String(d.Get("destination_cidr_block").(string)),
+		SourceCidrBlock:       aws.String(d.Get("source_cidr_block").(string)),
+		RuleAction:            aws.String(d.Get("rule_action").(string)),
+		RuleNumber:            aws.Int64(int64(d.Get("rule_number").(int))),
+		TrafficDirection:      aws.String(d.Get("traffic_direction").(string)),
+	}
+
+	if v, ok := d.GetOk("description"); ok {
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("protocol"); ok {
+		input.Protocol = aws.Int64(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("destination_port_range"); ok {
+		input.SetDestinationPortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	if v, ok := d.GetOk("source_port_range"); ok {
+		input.SetSourcePortRange(buildTrafficMirrorPortRangeRequest(v.([]interface{})))
+	}
+
+	out, err := conn.CreateTrafficMirrorFilterRule(input)
+	if err != nil {
+		return fmt.Errorf("Error creating traffic mirror filter rule for %v", filterId)
+	}
+
+	d.SetId(*out.TrafficMirrorFilterRule.TrafficMirrorFilterRuleId)
+	return resourceAwsEc2TrafficMirrorFilterRuleRead(d, meta)
+}
+
+func resourceAwsEc2TrafficMirrorFilterRuleRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+	ruleId := d.Id()
+
+	var rule *ec2.TrafficMirrorFilterRule
+	filterId, filterIdSet := d.GetOk("traffic_mirror_filter_id")
+	input := &ec2.DescribeTrafficMirrorFiltersInput{}
+	if filterIdSet {
+		input.TrafficMirrorFilterIds = []*string{aws.String(filterId.(string))}
+		awsMutexKV.Lock(filterId.(string))
+		defer awsMutexKV.Unlock(filterId.(string))
+	} else {
+		awsMutexKV.Lock("DescribeTrafficMirrorFilters")
+		defer awsMutexKV.Unlock("DescribeTrafficMirrorFilters")
+	}
+
+	found := false
+	for !found {
+		out, err := conn.DescribeTrafficMirrorFilters(input)
+		if err != nil {
+			return fmt.Errorf("Error listing traffic mirror filters in the account")
+		}
+
+		if 0 == len(out.TrafficMirrorFilters) {
+			return fmt.Errorf("No traffir mirror filters found")
+		}
+
+		_, rule, err = findRuleInFilters(ruleId, out.TrafficMirrorFilters)
+		if nil == err {
+			found = true
+			break
+		} else {
+			if out.NextToken == nil {
+				break
+			} else {
+				input.NextToken = out.NextToken
+			}
+		}
+	}
+
+	if !found {
+		d.SetId("")
+		return fmt.Errorf("Rule %s not found", ruleId)
+	}
+
+	return populateTrafficMirrorFilterRuleResource(d, rule)
```

Comment:
Since this function is only used once here, let's remove the indirection and include its `d.Set()`, etc. logic directly here. 👍